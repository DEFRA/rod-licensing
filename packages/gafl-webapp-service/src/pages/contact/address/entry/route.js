import { ADDRESS_ENTRY, ADDRESS_LOOKUP } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import { countries } from '../../../../processors/refdata-helper.js'
import { nextPage } from '../../../../routes/next-page.js'
import { getPronoun } from '../../../../processors/licence-type-display.js'

export const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  street: validation.contact.createStreetValidator(Joi),
  locality: validation.contact.createLocalityValidator(Joi),
  town: validation.contact.createTownValidator(Joi),
  postcode: Joi.alternatives().conditional('country-code', {
    is: Joi.string().pattern(/^GB/),
    then: validation.contact.createUKPostcodeValidator(Joi),
    otherwise: validation.contact.createOverseasPostcodeValidator(Joi)
  }),
  'country-code': Joi.string().required()
}).options({ abortEarly: false, allowUnknown: true })

export const getCountryDropDownOptions = async () => {
  const options = await countries.getAll()
  return options.filter(country => country.code !== 'GB')
}

export const getData = async request => {
  const { addresses, searchTerms } = await request.cache().helpers.addressLookup.getCurrentPermission()
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()

  const pronoun = getPronoun(isLicenceForYou).possessive

  return {
    pronoun,
    searchTerms: !addresses?.length && searchTerms ? searchTerms : null,
    countries: await getCountryDropDownOptions(),
    lookupPage: ADDRESS_LOOKUP.uri
  }
}

export default pageRoute(ADDRESS_ENTRY.page, ADDRESS_ENTRY.uri, validator, nextPage, getData)
