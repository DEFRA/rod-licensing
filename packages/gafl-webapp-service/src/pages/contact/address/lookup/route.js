import { ADDRESS_LOOKUP, ADDRESS_ENTRY, OS_TERMS } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { hasJunior, validation } from '@defra-fish/business-rules-lib'
import { isPhysical } from '../../../../processors/licence-type-display.js'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  postcode: validation.contact.createUKPostcodeValidator(Joi)
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    isLicenceForYou: permission.isLicenceForYou,
    licenceLength: permission.licenceLength,
    junior: hasJunior(permission),
    isPhysical: isPhysical(permission),
    uri: {
      entryPage: addLanguageCodeToUri(request, ADDRESS_ENTRY.uri),
      osTerms: addLanguageCodeToUri(request, OS_TERMS.uri)
    }
  }
}

export default pageRoute(ADDRESS_LOOKUP.page, ADDRESS_LOOKUP.uri, validator, nextPage, getData)
