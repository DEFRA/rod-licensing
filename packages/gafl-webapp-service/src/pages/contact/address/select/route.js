import { ADDRESS_SELECT, ADDRESS_LOOKUP, ADDRESS_ENTRY } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

export const getData = async request => {
  const { addresses, searchTerms } = await request.cache().helpers.addressLookup.getCurrentPermission()
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    isLicenceForYou,
    addresses,
    searchTerms,
    lookupPage: addLanguageCodeToUri(request, ADDRESS_LOOKUP.uri),
    entryPage: addLanguageCodeToUri(request, ADDRESS_ENTRY.uri)
  }
}

export default pageRoute(
  ADDRESS_SELECT.page,
  ADDRESS_SELECT.uri,
  Joi.object({
    address: Joi.number().integer().min(0).required()
  }).options({ abortEarly: false, allowUnknown: true }),
  nextPage,
  getData
)
