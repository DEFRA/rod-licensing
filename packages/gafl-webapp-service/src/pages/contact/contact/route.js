import { CONTACT, LICENCE_LENGTH, DATE_OF_BIRTH, LICENCE_TO_START } from '../../../uri.js'
import { HOW_CONTACTED } from '../../../processors/mapping-constants.js'
import pageRoute from '../../../routes/page-route.js'
import GetDataRedirect from '../../../handlers/get-data-redirect.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import { isPhysical } from '../../../processors/licence-type-display.js'
import { hasJunior } from '../../../processors/concession-helper.js'
import { nextPage } from '../../../routes/next-page.js'
import { mobilePhoneValidator } from '../../../processors/contact-validator.js'

const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  // We need to have set the licence length, dob and start date here to determining the contact
  // messaging
  if (!permission.licensee.birthDate) {
    throw new GetDataRedirect(DATE_OF_BIRTH.uri)
  }

  if (!permission.licenceStartDate) {
    throw new GetDataRedirect(LICENCE_TO_START.uri)
  }

  if (!permission.licenceLength) {
    throw new GetDataRedirect(LICENCE_LENGTH.uri)
  }

  return {
    licensee: permission.licensee,
    isPhysical: isPhysical(permission),
    isJunior: hasJunior(permission),
    howContacted: HOW_CONTACTED
  }
}

const validator = Joi.object({
  'how-contacted': Joi.string()
    .valid('email', 'text', 'none')
    .required(),
  email: Joi.alternatives().conditional('how-contacted', {
    is: 'email',
    then: validation.contact.createEmailValidator(Joi),
    otherwise: Joi.string().empty('')
  }),
  text: Joi.alternatives().conditional('how-contacted', {
    is: 'text',
    then: mobilePhoneValidator,
    otherwise: Joi.string().empty('')
  })
}).options({ abortEarly: false, allowUnknown: true })

export default pageRoute(CONTACT.page, CONTACT.uri, validator, nextPage, getData)
