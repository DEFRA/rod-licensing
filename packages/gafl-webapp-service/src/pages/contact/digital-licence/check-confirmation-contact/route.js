import { CHECK_CONFIRMATION_CONTACT, CONTACT, LICENCE_CONFIRMATION_METHOD } from '../../../../uri.js'
import { HOW_CONTACTED } from '../../../../processors/mapping-constants.js'
import GetDataRedirect from '../../../../handlers/get-data-redirect.js'
import pageRoute from '../../../../routes/page-route.js'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

export const getData = async request => {
  const { licensee } = await request.cache().helpers.transaction.getCurrentPermission()

  if (licensee.preferredMethodOfConfirmation === HOW_CONTACTED.none) {
    throw new GetDataRedirect(CONTACT.uri)
  }

  const whatToChange = licensee.preferredMethodOfConfirmation === HOW_CONTACTED.email ? '?change=email' : '?change=mobile'
  const change = addLanguageCodeToUri(request, `${LICENCE_CONFIRMATION_METHOD.uri}${whatToChange}`)

  return {
    licensee,
    uri: {
      contact: addLanguageCodeToUri(request, CONTACT.uri),
      change
    }
  }
}

export default pageRoute(CHECK_CONFIRMATION_CONTACT.page, CHECK_CONFIRMATION_CONTACT.uri, null, nextPage, getData)
