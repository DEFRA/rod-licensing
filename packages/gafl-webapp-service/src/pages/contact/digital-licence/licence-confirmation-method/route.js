import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'

import { CONTACT, LICENCE_CONFIRMATION_METHOD } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import GetDataRedirect from '../../../../handlers/get-data-redirect.js'
import { nextPage } from '../../../../routes/next-page.js'
import { HOW_CONTACTED } from '../../../../processors/mapping-constants.js'
import { isPhysical } from '../../../../processors/licence-type-display.js'
import { mobilePhoneValidator } from '../../../../processors/contact-validator.js'

const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  // page is only permitted for physical licences
  if (!isPhysical(permission)) {
    throw new GetDataRedirect(CONTACT.uri)
  }

  return {
    licensee: permission.licensee,
    howContacted: HOW_CONTACTED
  }
}

const validator = Joi.object({
  'licence-confirmation-method': Joi.string()
    .valid('email', 'text', 'none')
    .required(),
  text: Joi.alternatives().conditional('licence-confirmation-method', {
    is: 'text',
    then: mobilePhoneValidator,
    otherwise: Joi.string().empty('')
  }),
  email: Joi.alternatives().conditional('licence-confirmation-method', {
    is: 'email',
    then: validation.contact.createEmailValidator(Joi),
    otherwise: Joi.string().empty('')
  })
}).options({ abortEarly: false, allowUnknown: true })

export default pageRoute(LICENCE_CONFIRMATION_METHOD.page, LICENCE_CONFIRMATION_METHOD.uri, validator, nextPage, getData)
