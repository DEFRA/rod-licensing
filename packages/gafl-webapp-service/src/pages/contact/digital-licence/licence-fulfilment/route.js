import Joi from 'joi'

import pageRoute from '../../../../routes/page-route.js'
import { CONTACT, LICENCE_FULFILMENT } from '../../../../uri.js'
import { nextPage } from '../../../../routes/next-page.js'
import { isPhysical } from '../../../../processors/licence-type-display.js'
import GetDataRedirect from '../../../../handlers/get-data-redirect.js'

const isPhysicalPermission = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  // page is only permitted for physical licences
  if (!isPhysical(permission)) {
    throw new GetDataRedirect(CONTACT.uri)
  }
}

const validator = Joi.object({
  'licence-option': Joi.string()
    .valid('digital', 'paper-licence')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

export default pageRoute(LICENCE_FULFILMENT.page, LICENCE_FULFILMENT.uri, validator, nextPage, isPhysicalPermission)
