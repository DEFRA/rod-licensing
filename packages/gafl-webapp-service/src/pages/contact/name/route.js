import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { NAME } from '../../../uri.js'
import { validation } from '@defra-fish/business-rules-lib'
import { nextPage } from '../../../routes/next-page.js'

const validator = Joi.object({
  'first-name': validation.contact.createFirstNameValidator(Joi),
  'last-name': validation.contact.createLastNameValidator(Joi)
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()
  const permissions = await request.cache().helpers.transaction.getCurrentPermission()
  const firstName = permissions.licensee.firstName
  const lastName = permissions.licensee.lastName

  return {
    isLicenceForYou,
    firstName,
    lastName
  }
}

const namePageRoute = pageRoute(NAME.page, NAME.uri, validator, nextPage, getData)

// Sanitize does not play well with the name validator and is unnecessary
Object.assign(namePageRoute[1].options, {
  plugins: {
    disinfect: {
      disinfectQuery: true,
      disinfectParams: false,
      disinfectPayload: false
    }
  }
})

export { namePageRoute as default }
