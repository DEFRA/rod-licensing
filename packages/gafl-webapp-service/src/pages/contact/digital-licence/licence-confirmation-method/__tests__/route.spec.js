import { CONTACT } from '../../../../../uri.js'
import { getData } from '../route.js'
import { isPhysical } from '../../../../../processors/licence-type-display.js'

jest.mock('../../../../../processors/licence-type-display.js', () => ({
  isPhysical: jest.fn(() => true)
}))

describe('licence-confirmation-method > route', () => {
  describe('getData', () => {
    const getMockRequest = (options = {}) => ({
      cache: jest.fn(() => ({
        helpers: {
          transaction: {
            getCurrentPermission: jest.fn(() => ({
              licenceLength: options.licenceLength || '12M',
              licensee: {
                firstName: 'Lando',
                lastName: 'Norris'
              },
              isLicenceForYou: options.isLicenceForYou || false
            }))
          }
        }
      })),
      query: options.query || {}
    })

    beforeEach(jest.clearAllMocks)

    it('should reject and redirect to the contact page, if licence is not physical', async () => {
      isPhysical.mockReturnValueOnce(false)
      const func = async () => await getData(getMockRequest())
      await expect(func).rejects.toThrowRedirectTo(CONTACT.uri)
    })

    it('returns the expected data', async () => {
      expect(await getData(getMockRequest())).toMatchSnapshot()
    })

    describe('if change query is undefined,', () => {
      let result
      beforeEach(async () => {
        result = await getData(getMockRequest())
      })

      it("doesn't return changeEmail", async () => {
        expect(result.changeEmail).toBeUndefined()
      })

      it("doesn't return changeMobile", async () => {
        expect(result.changeMobile).toBeUndefined()
      })
    })

    it('if change query param is "email", returns changeEmail set to true', async () => {
      const result = await getData(getMockRequest({ query: { change: 'email' } }))
      expect(result.changeEmail).toBe(true)
    })

    it('if change query param is "mobile", returns changeMobile set to true', async () => {
      const result = await getData(getMockRequest({ query: { change: 'mobile' } }))
      expect(result.changeMobile).toBe(true)
    })

    it('should return isLicenceForYou as true, if isLicenceForYou is true on the transaction cache', async () => {
      const result = await getData(getMockRequest({ isLicenceForYou: true }))
      expect(result.isLicenceForYou).toBeTruthy()
    })

    it('should return isLicenceForYou as false, if isLicenceForYou is false on the transaction cache', async () => {
      const result = await getData(getMockRequest({ isLicenceForYou: false }))
      expect(result.isLicenceForYou).toBeFalsy()
    })
  })
})
