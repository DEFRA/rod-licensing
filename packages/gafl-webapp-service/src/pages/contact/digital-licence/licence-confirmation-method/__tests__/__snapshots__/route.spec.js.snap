// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`licence-confirmation-method > route getData returns the expected data 1`] = `
Object {
  "howContacted": Object {
    "email": "Email",
    "letter": "Letter",
    "none": "Prefer not to be contacted",
    "text": "Text",
  },
  "isLicenceForYou": false,
  "licensee": Object {
    "firstName": "Lando",
    "lastName": "Norris",
  },
}
`;
